core/src/main/scala/akka/kafka/scaladsl/Committer.scala

```diff
@@ -26,18 +28,48 @@ object Committer {
    */
   def batchFlow(settings: CommitterSettings): Flow[Committable, CommittableOffsetBatch, NotUsed] =
     Flow[Committable]
-    // Not very efficient, ideally we should merge offsets instead of grouping them
       .groupedWeightedWithin(settings.maxBatch, settings.maxInterval)(_.batchSize)
       .map(CommittableOffsetBatch.apply)
       .mapAsync(settings.parallelism) { b =>
         b.commitScaladsl().map(_ => b)(ExecutionContexts.sameThreadExecutionContext)
       }
 
+  /**
+   * API MAY CHANGE
+   *
+   * Batches offsets from context and commits them to Kafka, emits no useful value, but keeps the committed
+   * `CommittableOffsetBatch` as context.
+   */
+  @ApiMayChange
+  def flowWithOffsetContext[E](
+      settings: CommitterSettings
+  ): FlowWithContext[E, Committable, NotUsed, CommittableOffsetBatch, NotUsed] = {
+    val value: Flow[(E, Committable), (NotUsed, CommittableOffsetBatch), NotUsed] = Flow[(E, Committable)]
+      .map(_._2)
+      .groupedWeightedWithin(settings.maxBatch, settings.maxInterval)(_.batchSize)
+      .map(CommittableOffsetBatch.apply)
+      .mapAsync(settings.parallelism) { b =>
+        b.commitScaladsl().map(_ => (NotUsed, b))(ExecutionContexts.sameThreadExecutionContext)
+      }
+    new FlowWithContext(value)
+  }
+
   /**
    * Batches offsets and commits them to Kafka.
    */
   def sink(settings: CommitterSettings): Sink[Committable, Future[Done]] =
     flow(settings)
       .toMat(Sink.ignore)(Keep.right)
 
+  /**
+   * API MAY CHANGE
+   *
+   * Batches offsets from context and commits them to Kafka.
+   */
+  @ApiMayChange
+  def sinkWithOffsetContext[E](settings: CommitterSettings): Sink[(E, Committable), Future[Done]] =
```

Comment:
We still need to fix the name here. `sinkWithCommitableContext`? Same for the flow above?

Reply:
But it is the same as the flow?! `withOffsetContext`

Reply:
Yes. I think that one should also be renamed to `flowWithCommittableContext` as that is what the context type is.

Reply:
This is a hard one. The API is actually lying, it can't handle "any" `Committable` as everything but offsets will fail here

https://github.com/akka/alpakka-kafka/blob/master/core/src/main/scala/akka/kafka/internal/MessageBuilder.scala#L131-L134

Reply:
I changed those new `withOffsetContext` methods to require `CommittableOffset` other `Committable`s won't do anyway.